File: android-beans/src/main/java/rocks/frieler/android/beans/scopes/GenericScopedFactoryBean.java
package rocks.frieler.android.beans.scopes;

import java8.util.function.Supplier;

public class GenericScopedFactoryBean<T> implements ScopedFactoryBean<T> {
    private final String scope;
    private final Class<T> type;
    private final Supplier<T> producer;

    public GenericScopedFactoryBean(String scope, Class<T> type, Supplier<T> producer) {
        this.scope = scope;
        this.type = type;
        this.producer = producer;
    }

    @Override
    public String getScope() {
        return scope;
    }

    @Override
    public Class<T> getBeanType() {
        return type;
    }

    @Override
    public T produceBean() {
        return producer.get();
    }
}

File: android-beans/src/main/kotlin/rocks/frieler/android/beans/scopes/GenericScopedFactoryBean.kt
package rocks.frieler.android.beans.scopes

import rocks.frieler.android.beans.BeansProvider
import kotlin.reflect.KClass

open class GenericScopedFactoryBean<T : Any>(
        override val scope: String,
        override val beanType: KClass<T>,
        private val producer: BeansProvider.() -> T)
    : ScopedFactoryBean<T> {

    override fun produceBean(dependencies: BeansProvider): T {
        return producer(dependencies)
    }
}
